fms/models/__init__.py

```diff
@@ -82,4 +87,157 @@ def get_model(
+
+def load_model(
+    architecture: str,
+    variant: str,
+    model_path: str = None,
+    source: str = None,
+    device_type: str = "cpu",
+    distributed_strategy: str = None,
+    checkpoint_sharding: str = None,
+    group: ProcessGroup = None,
+    **kwargs,
+):
+    """
+    Load an instance of a model with weights.
+
+    Args:
+    architecture: the model architecture, e.g. llama. See
+                `models.list_models()`.
+    variant: the configuration of the model, e.g. 7b. See
+                `models.list_variants(architecture)`
+    model_path: the path to the state_dict of weights
+    device_type: where to load the model
+    distributed_strategy: None, 'fsdp', 'hsdp', 'tp', or 'mp'.
+    checkpoint_sharding: how the checkpoint files are sharded: None, 'tp',
+                'fsdp', or 'layer'. If None, guess based on files.
+    source: If the weights in the state dict didn't come from an FMS model,
+                `source` specifies which conversion function might be needed.
+                See `serialization.list_sources(architecture)`
+    group: ProcessGroup
+    """
+    local_rank, world_size = distributed.rank_and_world(group)
+
+    if distributed_strategy is None or distributed_strategy == "":
+        if world_size > 1:
+            distributed_strategy = "tp"
+        elif torch.cuda.device_count() > 1:
+            distributed_strategy = "mp"
+
+    device = torch.device(device_type, local_rank)
+    initial_device = device
+    if (
+        _is_dp(distributed_strategy)
+        and local_rank != 0
+        and checkpoint_sharding != "fsdp"
+    ):
+        initial_device = "meta"
+
+    fms_sd = serialization.load_state_dict(
+        model_path,
+        distributed_strategy,
+        checkpoint_sharding,
+        initial_device,
+        local_rank,
+        world_size,
+    )
+    fms_sd = serialization.get_adapted(architecture, source, fms_sd)
+
+    extra_args = kwargs
+    if distributed_strategy == "tp":
+        print("using tensor parallel")
+        extra_args["distributed_strategy"] = TensorParallelStrategy()
+    elif distributed_strategy == "mp":
+        print("using model parallel")
+        devices = [i for i in range(torch.cuda.device_count())]
+        extra_args["distributed_strategy"] = UniformModelParallelStrategy(
+            devices, _guess_num_layers(fms_sd)
+        )
+
+    ibm_model = get_model(
+        architecture, variant, device=initial_device, extra_args=extra_args
+    )
+
+    # In some cases we'd need to load the sd before distributing, in some cases
+    # after.
+    # e.g. a checkpoint can be loaded onto rank0 and synced across ranks by
+    # FSDP, but in most cases we need to load the rank-specific checkpoint
+    # onto the current rank.
+    pre_load = (
+        distributed_strategy in ["fsdp", "hsdp"] and checkpoint_sharding != "fsdp"
+    )
+
+    if pre_load and local_rank == 0 and fms_sd is not None:
+        ibm_model.load_state_dict(fms_sd, strict=False)
```

Comment:
is there any reason why strict=True cannot be done here?

Reply:
we never did with any of our previous adapters so I assumed it didn't work? if it does I'd rather test that change separately. distribution strategy could affect what layers are available where.